### ISSUE TITLE
Rate plans revisited

### ISSUE THREAD

**Issue by JirkaChadima** [windingtree/wiki#45]
In order to fulfill https://github.com/orgs/windingtree/projects/1#card-10605714 and as discussed in #28, we should re-visit the rate plans data model design and implement at least part of it.

The original state before this issue can be seen [here](https://github.com/windingtree/wiki/blob/57cae0e2c8952cc9f7b755fae292d153ba331e62/hotel-data-swagger.yaml#L184).

We should re-consider some things, namely:

- [x] cancellation terms
- [x] (relative) discounts
- [x] restrictions
- [x] rapidly changing prices

**Comment by JirkaChadima** [comment]
My ideas: 

*Cancellation*

- It should not be tied to any particular rate plan. It seems to me as a separate policy thing that applies to the whole hotel. But maybe we can do a relation between cancellation policy and a rate plan. I'd put it into description document. If it's not enought we can later make cancellation policies tied to rate plans directly.
- There should be multiple cancellation policies that would differ by a deadline (i. e. how many days prior to arrival does the given policy apply).
- We have to have a non refundable cancellation policy.
- A cancellation policy should probably be limited by dates (i. e. different policies for off-season months).
- It should be based on percentages.
- OTA standard has also guarantees and deposits. I wouldn't bother with them for now. The don't really seem related to cancellation policies to me.

Very rough proposal:
```
{
"cancellationPolicies": [
  {
    "start": "ISO8601", # when does this policy apply
    "end": "ISO8601",
    "deadline": "30", # how many days before arrival does this first apply (i. e. 30 or less inclusive)
    "amount": "25", # cost of cancellation in percents of the final price, 100 means non refundable
  }
]
}
```

*Discounts && Restrictions*

- The concept of discounts can be generalized into price modifiers that can be applied for certain dates and under certain conditions.
- Restrictions are affecting the daily price as well.
- I was thinking that we could simplify the rate plan structure by having only a single base price in it and move all restrictions, additional prices, length of stay based prices and other variables outside of the rate plan in a standalone structure.
- If you'd want to find out what price you should give to the customer, you would apply all (or only the best?) fitting modifiers.

```
{
  "rate-plan: {
    "name": "..",
    "description": "..",
    "currency": "..",
    "roomTypeIds": [],
    "updatedAt": "..",
    "availableForReservation": {"start": "", "end": ""},
    "availableForTravel": {"start": "", "end": ""},
    "price": 150, # base price
    "modifiers": [
      {
         "start": "..",
         "end": "..",
         "conditions": { # multiple conditions affecting the resulting price
             "lengthOfStay": {
                  "min": 3,
                  "max": 6
             }
         },
         "adjustment": -3 # Percents of the base price
      }
    ]
  }
}
```

*Rapid changes*

- We should be able to rapidly change the modifiers section. Maybe we can model it as a standalone document to save some cycles.

**Comment by JirkaChadima** [comment]
I've added the `cancellationPolicies` and adjusted the rate plan design in a 
[new version](https://app.swaggerhub.com/apis/windingtree/WTHotelDefinitions/rateplans-2-draft) of the data structure specification. Please, have a look at it @czervenka and if it makes sense to you, publish it and copy it over to this repository via #47 (probably with a better version number).

However, I'm still not exactly happy with it, it's a really complex topic with lots of variables. I think that without some real-life usage, we are unable to design a one-size-fits-all solution. What we certainly need to do is to communicate well that this is *not* the perfect data format and the data itself might need to be reshaped to perform some operations (i. e. a reseller might want to pre-compute a price matrix for each date and room and restrictions and modifiers). We also have to clearly communicate which side of the chain are we targeting - are we trying to ease adoption of this for hotels, or are we trying to make the data consumption easy? We are doing the former, by the way.

For some pricing models, this data structure is probably totally inefficient. Maybe we need to forget about existing systems using OTA standard and create a brand new data format. But we really can't make that decision without some experience with real-life data. Finding the balance between ease of adoption and the ability to support more recent pricing models is really hard.

But for now, this is what we have and we will implement it in the wt-js-libs, wt-read-api and wt-write-api in its entirety for the sumer release. I'll prepare PRs in all three repositories so ti should be easy to reflect eventual changes that will happen in #47

**Comment by hynek-urban** [comment]
@JirkaChadima My uninformed opinion (I wasn't present at the meetings where you've probably discussed this - so perhaps I'm mistaken) - the sketched concept of "price modifiers" seems potentially very complicated. 

I'm talking about:

1. The principle - to get an actual price, you have to have a good understanding of the data and perform a number of operations to select and apply the proper modifiers. I foresee this could be complicated for data consumers and prone to errors from data producers (e.g. mistakenly create a combination of price modifiers that results in an unprofitable price plan) .

2. I suspect the domain of possible conditions and operators in the "conditions" could grow substantially, basically forcing users of this data model to learn a dedicated DSL just for rate plans.

**Comment by hynek-urban** [comment]
After my work on wt-fixtures, where I've tried to prepare data "as if" I was a manager of different types of hotels, I have stumbled upon the following additional problems (all come from situations that I have seen in the wild):

1. I have a room, whose price differs based on how many people occupy it. Example: three-bed room costing 40 EUR per night for a single person, 60 EUR for two people and 75 for three people. I think this is a rather common case that currently cannot be encoded in a reasonable way, neither using "RatePlan.restrictions", nor "RatePlan.modifiers".

2. I would suggest making "availableForReservation" and "availableForTravel" arrays of date intervals. Otherwise, various pricings have to be encoded redundantly using multiple rate plans even though they are effectively the same (e.g. out of season prices before and after the high season).

3. What if I only rent rooms/apartments for whole weeks? What if the price is not per night but per week? (Nothing in the spec says it's per night.)

4. What about shared rooms / dorms (hostels)? How can I capture the fact that only part of the room will be effectively booked?

Additionaly:

5. The cancellation policy deadline description seems a bit misleading to me: "How many days before arrival does this first apply." I suspect the cancellation policy actually stops being applicable on this day.

**Comment by czervenka** [comment]
Good feedback from pseudo real-life situation

ad 1) You are right, would minOccupants condition in ratePlanModifier solve that?
ad 2) Maybe, but I would wait for real life issues. WT is not meant to be a normalised database.
ad 3) Let's add this to todo list (#28) ... Let's add to the description that the prices are per night (if it is not there yet).
ad 4) In case of hostel bed=room. In other words when you book a room, you booked a bed in a room. Room would probably equals to roomType in this case.

**Comment by hynek-urban** [comment]
@czervenka 

Thanks for the reply. 

1) I'm not sure. It might rather be maxOccupants, I don't know - depends on how the ratePlan resolution actually works :)
2) My motivation was the following: having a hotel with 5 room types, two high-season (e.g. summer and winter holiday seasons) and two off-season periods means I have to create 5 * 4 = 20 ratePlans instead of 10. That's quite a PITA. But perhaps you're right.
4) For some reason, I have rejected this alternative previously, but I can't recall why now :) So this might actually work.

**Comment by hynek-urban** [comment]
There's another issue I've just stumbled across: the "maxAge" price modifier. What if there's a mixed occupancy (typically, a parent and a child); how does it apply? The description says " The price is applicable to occupants of this age or younger ". However, we generally do not have "per person" prices but "per room ones" so I'm not sure how it would be computed in this case.

**Comment by JirkaChadima** [comment]
Good feedback, @hynek-urban ! I think a lot of the confusion comes from a lack of documentation :-)

> I have a room, whose price differs based on how many people occupy it.

This was captured in the original rate plans design, but I'm pretty sure it can be done with price modifiers as @czervenka suggests. I guess this slipped when I was changing the spec. I agree that this is something we have to support.

> I would suggest making "availableForReservation" and "availableForTravel" arrays of date intervals. 

This really depends on how the originating systems are behaving. If they are flushing the data to WT in isolated bursts, it might be complicated for them to collect all the data necessary to give us more than one date ranges. I guess we have to make an opinionated decision on this (and some other things) at some point and decide how much we want to make it easy for existing systems to transform their data in our format.

> My motivation was the following: having a hotel with 5 room types, two high-season...

This again depends on who's actually giving us the data. If you make it by hand, it's PITA. If a system is doing that for you, no one cares. The only issue I can see with this is potentially unnecessary large size of the data. Also this is along the line with the decision on who we want to be close to: Data providers or data consumers? I don't know.

> What if I only rent rooms/apartments for whole weeks?

That's a good idea, I'd add it to #28. Although you can model it with prices per night.

> It might be useful to expand the cancellation policy description

How exactly? Also, I'd rather not introduce implicit defaults.

> There's another issue I've just stumbled across: the "maxAge" price modifier. What if there's a mixed age occupancy (typically, a parent and a child); how does it apply? The description says " The price is applicable to occupants of this age or younger ". However, we generally do not have "per person" prices but "per room" ones so I'm not sure how it would be computed in this case.

OK, the documentation probably sucks :-) But in the RatePlan section there's a very crude total price computing algorithm, that says `for every person, compute the price` or something like that. So I believe our (or at leasy my) thinking is that the stated price is **per room per night per person**.

We will probably get a lot of similar feedback once someone actually starts using it, but we should somehow somewhere draw the line for summer release and incorporate all the ideas later. This won't be perfect the first time.

**Comment by JirkaChadima** [comment]
OK, so to conclude this iteration:

**FIX - see 7d30f58**

- We have to add a mandatory `defaultCancellationAmount` field next to `cancellationPolicies`
- `from` and `to` in RatePlanModifer are moved to `conditions` object. They are just another set of conditions.
- We have to clearly document how are the RatePlanModifiers combined together.
- I've added `minOccupants` into RatePlanModifier conditions.

**TODO (added to #28)**

- Add more RatePlanModifier resolution strategies - after all, there are modifiers that might raise the price for legitimate reasons (for example pets)
- Cancellation deadline should have more granularity than days
- Some pricing models might not work with prices per night per room per person.
- We need to improve the list of available modifiers and restrictions - see Occupancy based pricing.
- It might make sense to "group" rate plans together (for `availableForTravel` and `availableForReservation` fields)
